001-clarifications-eval.txt

```diff
@@ -0,0 +1,754 @@
+Part 1: Clarifications for Expression Evaluation
+
+In Section 2.5 DWARF Expressions, add the following text after the first
+paragraph:
+
+    --------------------------------------------------------------------
+    [non-normative] The evaluation of a DWARF expression can provide the
+    location of an object, the value of an array bound, the length of a
+    dynamic string, the desired value itself, and so on.
+
+    If the evaluation of a DWARF expression does not encounter an error,
+    then it can either result in a value (see 2.5.2 DWARF Expression
+    Value) or a location description (see 2.5.3 DWARF Location
+    Description). When a DWARF expression is evaluated, it may be
+    specified whether a value or location description is required as the
+    result kind.
+
+    If a result kind is specified, and the result of the evaluation does
+    not match the specified result kind, then the implicit conversions
+    described in 2.5.4.4.3 Memory Location Description Operations are
+    performed if valid. Otherwise, the DWARF expression is ill-formed.
+
+    If the evaluation of a DWARF expression encounters an evaluation
+    error, then the result is an evaluation error.
+
+    If a DWARF expression is ill-formed, then the result is undefined.
+
+    The following sections detail the rules for when a DWARF expression
+    is ill-formed or results in an evaluation error.
+    --------------------------------------------------------------------
+
+Insert the following subsection under 2.5 (renumbering subsequent
+subsections):
+
+    --------------------------------------------------------------------
+    2.5.1 DWARF Expression Evaluation Context
+
+    A DWARF expression is evaluated in a context that can include a
+    number of context elements. If multiple context elements are
+    specified then they must be self consistent or the result of the
+    evaluation is undefined. The context elements that can be specified
+    are:
+
+    1. A current result kind
+
+        The kind of result required by the DWARF expression evaluation.
+        If specified it can be a location description or a value.
+
+    2. A current thread
+
+        The target architecture thread identifier of the source program
+        thread of execution for which a user presented expression is
+        currently being evaluated.
+
+        It is required for operations that are related to target
+        architecture threads.
+
+        [non-normative] For example: the DW_OP_regval_type operation.
+
+    3. A current call frame
+
+        The target architecture call frame identifier. It identifies a
```

Comment:
A call frame does have a Canonical Frame Address. However, during the execution of a program, the same CFA address could refer to different call frames as frames come and go. The intent of the context is to refer to the conceptual call frame and leave it to the implementation for how it wants to represent that. For example, in gdb there is a data type for call frames that it uses which is not the CFA. Hense this is deliberately referring to the identity of a call frame, without madating how to identify it.

Reply:
I think the intent is to refer to the call frame as being architecture specific. Then to introduce the concept that a call frame can be identified. In the evaluation context it is the identity of the call frame that is recorded. Later the notion of a Canonical Frame Address is introduced and it is defined how it relates to the identity of a call frame.

If CFA was used here, then it would need the later definition to be pulled up. I think it is better to leave this text using the more abstract concept of the identity of a target architecture call frame. And leave the introduction of a CFA concept (which builds on the concept of the identity of a call frame) until later.

Reply:
>If CFA was used here

CFA is used here two sentences later.

Reply:
I guess the "later" was nearer than I thought:-) How do you suggest wording this?